Build failure: z3-4.8.5

### Steps To Reproduce

Steps to reproduce the behavior:
1. build z3.4.8.5 using `nixos-unstable`

### Build log

`nix build .#z3_4_8_5`

```
z3> Traceback (most recent call last):
z3>   File "/build/source/scripts/mk_make.py", line 9, in <module>
z3>     from mk_util import *
z3>   File "/build/source/scripts/mk_util.py", line 18, in <module>
z3>     import distutils.sysconfig
z3> ModuleNotFoundError: No module named 'distutils'
error: builder for '/nix/store/rksx05p231nvzb0icz3hf6fb8hxhik2b-z3-4.8.5.drv' failed with exit code 1;
       last 10 log lines:
       > /build/source/scripts/mk_genfile_common.py:764: SyntaxWarning: invalid escape sequence '\('
       >   initializer_prio_pat = re.compile('[ \t]*ADD_INITIALIZER\(\'([^\']*)\',[ \t]*(-?[0-9]*)\)')
       > /build/source/scripts/mk_genfile_common.py:765: SyntaxWarning: invalid escape sequence '\('
       >   finalizer_pat        = re.compile('[ \t]*ADD_FINALIZER\(\'([^\']*)\'\)')
       > Traceback (most recent call last):
       >   File "/build/source/scripts/mk_make.py", line 9, in <module>
       >     from mk_util import *
       >   File "/build/source/scripts/mk_util.py", line 18, in <module>
       >     import distutils.sysconfig
       > ModuleNotFoundError: No module named 'distutils'
       For full logs, run 'nix log /nix/store/rksx05p231nvzb0icz3hf6fb8hxhik2b-z3-4.8.5.drv'.
```

### Additional context

`distutil` removal in Python 3.12 that is now the default Python version since #320924.
z3 4.8.5 is an `fstar` package dependency.
Other versions of z3 work, I will look for a possible fix.

### Notify maintainers

@thoughtpolice @ttuegel 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

Looks like the python package `setuptools` needs to be installed for Python 3.12 and also imported with `import setuptools` before `import distutils` wherever `distutils` was being used. I am still quite a novice when it comes to the inner workings of NixOS but the below should hopefully help pinpoint where changes need to be made

```
@nix { "action": "setPhase", "phase": "checkPhase" }
Running phase: checkPhase
check flags: -j16 check-runtimes
[0/1] Running runtimes regression tests
/build/source/compiler-rt/test/lit.common.cfg.py:639: SyntaxWarning: invalid escape sequence '\$'
  config.substitutions.append( ("%ld_flags_rpath_exe" + postfix, "-Wl,-z,origin -Wl,-rpath,\$ORIGIN -L%T -l%xdynamiclib_namespec" + postfix) )
/build/source/compiler-rt/test/lit.common.cfg.py:642: SyntaxWarning: invalid escape sequence '\$'
  config.substitutions.append( ("%ld_flags_rpath_exe" + postfix, "-Wl,-rpath,\$ORIGIN -L%T -l%xdynamiclib_namespec" + postfix) )
/build/source/compiler-rt/test/lit.common.cfg.py:645: SyntaxWarning: invalid escape sequence '\$'
  config.substitutions.append( ("%ld_flags_rpath_exe" + postfix, "-Wl,-R\$ORIGIN -L%T -l%xdynamiclib_namespec" + postfix) )
lit: /build/source/compiler-rt/test/lit.common.cfg.py:148: warning: Compiler lib dir != compiler-rt lib dir
Compiler libdir:     "/nix/store/kjav83vr3vhq96hj4wr95fbkxdgwcbqg-rocm-llvm-runtimes-6.0.2/lib/linux"
compiler-rt libdir:  "/build/source/runtimes/build/compiler-rt/lib/linux"
lit: /build/source/compiler-rt/test/lit.common.cfg.py:158: warning: COMPILER_RT_TEST_STANDALONE_BUILD_LIBS=ON, but this test suite does not support testing the just-built runtime libraries when the test compiler is configured to use different runtime libraries. Either modify this test suite to support this test configuration, or set COMPILER_RT_TEST_STANDALONE_BUILD_LIBS=OFF to test the runtime libraries included in the compiler instead.
lit: /build/source/compiler-rt/test/lit.common.cfg.py:169: note: Testing using libraries in "/build/source/runtimes/build/compiler-rt/lib/linux"
lit: /nix/store/yz7yczcfpipg2gdbs5xm9pxxl84h1pq7-python3.12-lit-18.1.7/lib/python3.12/site-packages/lit/TestingConfig.py:152: fatal: unable to parse config file '/build/source/compiler-rt/test/lit.common.cfg.py', traceback: Traceback (most recent call last):
  File "/nix/store/yz7yczcfpipg2gdbs5xm9pxxl84h1pq7-python3.12-lit-18.1.7/lib/python3.12/site-packages/lit/TestingConfig.py", line 140, in load_from_path
    exec(compile(data, path, "exec"), cfg_globals, None)
  File "/build/source/compiler-rt/test/lit.common.cfg.py", line 519, in <module>
    from distutils.version import LooseVersion
ModuleNotFoundError: No module named 'distutils'

FAILED: CMakeFiles/check-runtimes /build/source/runtimes/build/CMakeFiles/check-runtimes 
cd /build/source/runtimes/build && /nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4/bin/python3.12 /nix/store/yz7yczcfpipg2gdbs5xm9pxxl84h1pq7-python3.12-lit-18.1.7/bin/.lit-wrapped -sv --show-xfail --show-unsupported /build/source/runtimes/build/compiler-rt/test/builtins/TestCases /build/source/runtimes/build/compiler-rt/test/builtins/Unit/X86_64LinuxConfig
ninja: build stopped: subcommand failed.
```
This is where is fails for me, I used `nix log /nix/store/x0cc4afsyk5m2ly5lhg2mycwqbg2zarb-rocm-llvm-compiler-rt-6.0.2.drv` to get this output.

--- a/nixos/doc/manual/release-notes/rl-2411.section.md
+++ b/nixos/doc/manual/release-notes/rl-2411.section.md
@@ -928,6 +928,8 @@
 
 - `buildNimSbom` was added as an alternative to `buildNimPackage`. `buildNimSbom` uses [SBOMs](https://cyclonedx.org/) to generate packages whereas `buildNimPackage` uses a custom JSON lockfile format.
 
+- The default version of `z3` has been updated from 4.8 to 4.13. There are still a few packages that need specific 4.8 versions; those will continue to be maintained.
+
 ## Detailed Migration Information {#sec-release-24.11-migration}
 
 ### `sound` options removal {#sec-release-24.11-migration-sound}

--- a/pkgs/applications/science/logic/z3/4-8-5-typos.diff
+++ b/pkgs/applications/science/logic/z3/4-8-5-typos.diff
@@ -1,0 +1,26 @@
+diff --git a/src/util/lp/lp_core_solver_base.h b/src/util/lp/lp_core_solver_base.h
+index 4c17df2..4c3c311 100644
+--- a/src/util/lp/lp_core_solver_base.h
++++ b/src/util/lp/lp_core_solver_base.h
+@@ -600,8 +600,6 @@ public:
+            out << " \n";
+     }
+ 
+-    bool column_is_free(unsigned j) const { return this->m_column_type[j] == free; }
+-
+     bool column_has_upper_bound(unsigned j) const {
+         switch(m_column_types[j]) {
+         case column_type::free_column:
+diff --git a/src/util/lp/static_matrix_def.h b/src/util/lp/static_matrix_def.h
+index 7949573..2f1cb42 100644
+--- a/src/util/lp/static_matrix_def.h
++++ b/src/util/lp/static_matrix_def.h
+@@ -86,7 +86,7 @@ static_matrix<T, X>::static_matrix(static_matrix const &A, unsigned * /* basis *
+     init_row_columns(m, m);
+     while (m--) {
+         for (auto & col : A.m_columns[m]){
+-            set(col.var(), m, A.get_value_of_column_cell(col));
++            set(col.var(), m, A.get_column_cell(col));
+         }
+     }
+ }

--- a/pkgs/applications/science/logic/z3/default.nix
+++ b/pkgs/applications/science/logic/z3/default.nix
@@ -95,8 +95,8 @@ let common = { version, sha256, patches ? [ ], tag ? "z3" }:
 in
 {
   z3_4_12 = common {
-    version = "4.12.5";
-    sha256 = "sha256-Qj9w5s02OSMQ2qA7HG7xNqQGaUacA1d4zbOHynq5k+A=";
+    version = "4.12.6";
+    sha256 = "sha256-X4wfPWVSswENV0zXJp/5u9SQwGJWocLKJ/CNv57Bt+E=";
   };
   z3_4_11 = common {
     version = "4.11.2";

--- a/pkgs/applications/science/logic/z3/default.nix
+++ b/pkgs/applications/science/logic/z3/default.nix
@@ -89,7 +89,7 @@ let common = { version, sha256, patches ? [ ], tag ? "z3" }:
       changelog = "https://github.com/Z3Prover/z3/releases/tag/z3-${version}";
       license = licenses.mit;
       platforms = platforms.unix;
-      maintainers = with maintainers; [ thoughtpolice ttuegel ];
+      maintainers = with maintainers; [ thoughtpolice ttuegel numinit ];
     };
   };
 in

--- a/pkgs/applications/science/logic/z3/default.nix
+++ b/pkgs/applications/science/logic/z3/default.nix
@@ -41,6 +41,13 @@ let common = { version, sha256, patches ? [ ], tag ? "z3" }:
     postPatch = lib.optionalString ocamlBindings ''
       export OCAMLFIND_DESTDIR=$ocaml/lib/ocaml/${ocaml.version}/site-lib
       mkdir -p $OCAMLFIND_DESTDIR/stublibs
+    '' + lib.optionalString ((lib.versionAtLeast python.version "3.12") && (lib.versionOlder version "4.8.14")) ''
+      # See https://github.com/Z3Prover/z3/pull/5729. This is a specialization of this patch for 4.8.5.
+      for file in scripts/mk_util.py src/api/python/CMakeLists.txt; do
+        substituteInPlace "$file" \
+          --replace-fail "distutils.sysconfig.get_python_lib()" "sysconfig.get_path('purelib')" \
+          --replace-fail "distutils.sysconfig" "sysconfig"
+      done
     '';
 
     configurePhase = lib.concatStringsSep " "

--- a/pkgs/applications/science/logic/z3/default.nix
+++ b/pkgs/applications/science/logic/z3/default.nix
@@ -16,7 +16,7 @@
 assert javaBindings -> jdk != null;
 assert ocamlBindings -> ocaml != null && findlib != null && zarith != null;
 
-let common = { version, sha256, patches ? [ ], tag ? "z3" }:
+let common = { version, sha256, patches ? [ ], tag ? "z3", doCheck ? true }:
   stdenv.mkDerivation rec {
     pname = "z3";
     inherit version sha256 patches;
@@ -58,9 +58,9 @@ let common = { version, sha256, patches ? [ ], tag ? "z3" }:
           ++ lib.optional pythonBindings "--python --pypkgdir=$out/${python.sitePackages}"
       ) + "\n" + "cd build";
 
-    doCheck = true;
+    inherit doCheck;
     checkPhase = ''
-      make test
+      make -j $NIX_BUILD_CORES test
       ./test-z3 -a
     '';
 
@@ -94,6 +94,10 @@ let common = { version, sha256, patches ? [ ], tag ? "z3" }:
   };
 in
 {
+  z3_4_13 = common {
+    version = "4.13.4";
+    sha256 = "sha256-8hWXCr6IuNVKkOegEmWooo5jkdmln9nU7wI8T882BSE=";
+  };
   z3_4_12 = common {
     version = "4.12.6";
     sha256 = "sha256-X4wfPWVSswENV0zXJp/5u9SQwGJWocLKJ/CNv57Bt+E=";

--- a/pkgs/applications/science/logic/z3/default.nix
+++ b/pkgs/applications/science/logic/z3/default.nix
@@ -2,6 +2,7 @@
   lib,
   stdenv,
   fetchFromGitHub,
+  fetchpatch,
   python,
   fixDarwinDylibNames,
   javaBindings ? false,
@@ -12,6 +13,7 @@
   findlib ? null,
   zarith ? null,
   writeScript,
+  replaceVars,
 }:
 
 assert javaBindings -> jdk != null;
@@ -123,6 +125,27 @@ let
         ];
       };
     };
+
+  static-matrix-def-patch = fetchpatch {
+    # clang / gcc fixes. fixes typos in some member names
+    name = "gcc-15-fixes.patch";
+    url = "https://github.com/Z3Prover/z3/commit/2ce89e5f491fa817d02d8fdce8c62798beab258b.patch";
+    includes = [ "src/math/lp/static_matrix_def.h" ];
+    hash = "sha256-rEH+UzylzyhBdtx65uf8QYj5xwuXOyG6bV/4jgKkXGo=";
+  };
+
+  static-matrix-patch = fetchpatch {
+    # clang / gcc fixes. fixes typos in some member names
+    name = "gcc-15-fixes.patch";
+    url = "https://github.com/Z3Prover/z3/commit/2ce89e5f491fa817d02d8fdce8c62798beab258b.patch";
+    includes = [ "src/@dir@/lp/static_matrix.h" ];
+    stripLen = 3;
+    extraPrefix = "src/@dir@/";
+    hash = "sha256-+H1/VJPyI0yq4M/61ay8SRCa6OaoJ/5i+I3zVTAPUVo=";
+  };
+
+  # replace @dir@ in the path of the given list of patches
+  fixupPatches = dir: map (patch: replaceVars patch { dir = dir; });
 in
 {
   z3_4_13 = common {
@@ -132,18 +155,53 @@ in
   z3_4_12 = common {
     version = "4.12.6";
     sha256 = "sha256-X4wfPWVSswENV0zXJp/5u9SQwGJWocLKJ/CNv57Bt+E=";
+    patches =
+      fixupPatches "math" [
+        ./lower-bound-typo.diff
+        static-matrix-patch
+      ]
+      ++ [
+        static-matrix-def-patch
+      ];
   };
   z3_4_11 = common {
     version = "4.11.2";
     sha256 = "sha256-OO0wtCvSKwGxnKvu+AfXe4mEzv4nofa7A00BjX+KVjc=";
+    patches =
+      fixupPatches "math" [
+        ./lower-bound-typo.diff
+        static-matrix-patch
+        ./tail-matrix.diff
+      ]
+      ++ [
+        static-matrix-def-patch
+      ];
   };
   z3_4_8 = common {
     version = "4.8.17";
     sha256 = "sha256-BSwjgOU9EgCcm18Zx0P9mnoPc9ZeYsJwEu0ffnACa+8=";
+    patches =
+      fixupPatches "math" [
+        ./lower-bound-typo.diff
+        static-matrix-patch
+        ./tail-matrix.diff
+      ]
+      ++ [
+        static-matrix-def-patch
+      ];
   };
   z3_4_8_5 = common {
     tag = "Z3";
     version = "4.8.5";
     sha256 = "sha256-ytG5O9HczbIVJAiIGZfUXC/MuYH7d7yLApaeTRlKXoc=";
+    patches =
+      fixupPatches "util" [
+        ./lower-bound-typo.diff
+        static-matrix-patch
+        ./tail-matrix.diff
+      ]
+      ++ [
+        ./4-8-5-typos.diff
+      ];
   };
 }

--- a/pkgs/applications/science/logic/z3/default.nix
+++ b/pkgs/applications/science/logic/z3/default.nix
@@ -1,97 +1,128 @@
-{ lib
-, stdenv
-, fetchFromGitHub
-, python
-, fixDarwinDylibNames
-, javaBindings ? false
-, ocamlBindings ? false
-, pythonBindings ? true
-, jdk ? null
-, ocaml ? null
-, findlib ? null
-, zarith ? null
-, writeScript
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  python,
+  fixDarwinDylibNames,
+  javaBindings ? false,
+  ocamlBindings ? false,
+  pythonBindings ? true,
+  jdk ? null,
+  ocaml ? null,
+  findlib ? null,
+  zarith ? null,
+  writeScript,
 }:
 
 assert javaBindings -> jdk != null;
 assert ocamlBindings -> ocaml != null && findlib != null && zarith != null;
 
-let common = { version, sha256, patches ? [ ], tag ? "z3", doCheck ? true }:
-  stdenv.mkDerivation rec {
-    pname = "z3";
-    inherit version sha256 patches;
-    src = fetchFromGitHub {
-      owner = "Z3Prover";
-      repo = "z3";
-      rev = "${tag}-${version}";
-      sha256 = sha256;
-    };
+let
+  common =
+    {
+      version,
+      sha256,
+      patches ? [ ],
+      tag ? "z3",
+      doCheck ? true,
+    }:
+    stdenv.mkDerivation rec {
+      pname = "z3";
+      inherit version sha256 patches;
+      src = fetchFromGitHub {
+        owner = "Z3Prover";
+        repo = "z3";
+        rev = "${tag}-${version}";
+        sha256 = sha256;
+      };
 
-    strictDeps = true;
+      strictDeps = true;
 
-    nativeBuildInputs = [ python ]
-      ++ lib.optional stdenv.hostPlatform.isDarwin fixDarwinDylibNames
-      ++ lib.optional javaBindings jdk
-      ++ lib.optionals ocamlBindings [ ocaml findlib ]
-    ;
-    propagatedBuildInputs = [ python.pkgs.setuptools ]
-      ++ lib.optionals ocamlBindings [ zarith ];
-    enableParallelBuilding = true;
+      nativeBuildInputs =
+        [ python ]
+        ++ lib.optional stdenv.hostPlatform.isDarwin fixDarwinDylibNames
+        ++ lib.optional javaBindings jdk
+        ++ lib.optionals ocamlBindings [
+          ocaml
+          findlib
+        ];
+      propagatedBuildInputs = [ python.pkgs.setuptools ] ++ lib.optionals ocamlBindings [ zarith ];
+      enableParallelBuilding = true;
 
-    postPatch = lib.optionalString ocamlBindings ''
-      export OCAMLFIND_DESTDIR=$ocaml/lib/ocaml/${ocaml.version}/site-lib
-      mkdir -p $OCAMLFIND_DESTDIR/stublibs
-    '' + lib.optionalString ((lib.versionAtLeast python.version "3.12") && (lib.versionOlder version "4.8.14")) ''
-      # See https://github.com/Z3Prover/z3/pull/5729. This is a specialization of this patch for 4.8.5.
-      for file in scripts/mk_util.py src/api/python/CMakeLists.txt; do
-        substituteInPlace "$file" \
-          --replace-fail "distutils.sysconfig.get_python_lib()" "sysconfig.get_path('purelib')" \
-          --replace-fail "distutils.sysconfig" "sysconfig"
-      done
-    '';
+      postPatch =
+        lib.optionalString ocamlBindings ''
+          export OCAMLFIND_DESTDIR=$ocaml/lib/ocaml/${ocaml.version}/site-lib
+          mkdir -p $OCAMLFIND_DESTDIR/stublibs
+        ''
+        +
+          lib.optionalString
+            ((lib.versionAtLeast python.version "3.12") && (lib.versionOlder version "4.8.14"))
+            ''
+              # See https://github.com/Z3Prover/z3/pull/5729. This is a specialization of this patch for 4.8.5.
+              for file in scripts/mk_util.py src/api/python/CMakeLists.txt; do
+                substituteInPlace "$file" \
+                  --replace-fail "distutils.sysconfig.get_python_lib()" "sysconfig.get_path('purelib')" \
+                  --replace-fail "distutils.sysconfig" "sysconfig"
+              done
+            '';
 
-    configurePhase = lib.concatStringsSep " "
-      (
-        [ "${python.pythonOnBuildForHost.interpreter} scripts/mk_make.py --prefix=$out" ]
+      configurePhase =
+        lib.concatStringsSep " " (
+          [ "${python.pythonOnBuildForHost.interpreter} scripts/mk_make.py --prefix=$out" ]
           ++ lib.optional javaBindings "--java"
           ++ lib.optional ocamlBindings "--ml"
           ++ lib.optional pythonBindings "--python --pypkgdir=$out/${python.sitePackages}"
-      ) + "\n" + "cd build";
+        )
+        + "\n"
+        + "cd build";
 
-    inherit doCheck;
-    checkPhase = ''
-      make -j $NIX_BUILD_CORES test
-      ./test-z3 -a
-    '';
+      inherit doCheck;
+      checkPhase = ''
+        make -j $NIX_BUILD_CORES test
+        ./test-z3 -a
+      '';
 
-    postInstall = ''
-      mkdir -p $dev $lib
-      mv $out/lib $lib/lib
-      mv $out/include $dev/include
-    '' + lib.optionalString pythonBindings ''
-      mkdir -p $python/lib
-      mv $lib/lib/python* $python/lib/
-      ln -sf $lib/lib/libz3${stdenv.hostPlatform.extensions.sharedLibrary} $python/${python.sitePackages}/z3/lib/libz3${stdenv.hostPlatform.extensions.sharedLibrary}
-    '' + lib.optionalString javaBindings ''
-      mkdir -p $java/share/java
-      mv com.microsoft.z3.jar $java/share/java
-      moveToOutput "lib/libz3java.${stdenv.hostPlatform.extensions.sharedLibrary}" "$java"
-    '';
+      postInstall =
+        ''
+          mkdir -p $dev $lib
+          mv $out/lib $lib/lib
+          mv $out/include $dev/include
+        ''
+        + lib.optionalString pythonBindings ''
+          mkdir -p $python/lib
+          mv $lib/lib/python* $python/lib/
+          ln -sf $lib/lib/libz3${stdenv.hostPlatform.extensions.sharedLibrary} $python/${python.sitePackages}/z3/lib/libz3${stdenv.hostPlatform.extensions.sharedLibrary}
+        ''
+        + lib.optionalString javaBindings ''
+          mkdir -p $java/share/java
+          mv com.microsoft.z3.jar $java/share/java
+          moveToOutput "lib/libz3java.${stdenv.hostPlatform.extensions.sharedLibrary}" "$java"
+        '';
 
-    outputs = [ "out" "lib" "dev" "python" ]
-      ++ lib.optional javaBindings "java"
-      ++ lib.optional ocamlBindings "ocaml";
+      outputs =
+        [
+          "out"
+          "lib"
+          "dev"
+          "python"
+        ]
+        ++ lib.optional javaBindings "java"
+        ++ lib.optional ocamlBindings "ocaml";
 
-    meta = with lib; {
-      description = "High-performance theorem prover and SMT solver";
-      mainProgram = "z3";
-      homepage = "https://github.com/Z3Prover/z3";
-      changelog = "https://github.com/Z3Prover/z3/releases/tag/z3-${version}";
-      license = licenses.mit;
-      platforms = platforms.unix;
-      maintainers = with maintainers; [ thoughtpolice ttuegel numinit ];
+      meta = with lib; {
+        description = "High-performance theorem prover and SMT solver";
+        mainProgram = "z3";
+        homepage = "https://github.com/Z3Prover/z3";
+        changelog = "https://github.com/Z3Prover/z3/releases/tag/z3-${version}";
+        license = licenses.mit;
+        platforms = platforms.unix;
+        maintainers = with maintainers; [
+          thoughtpolice
+          ttuegel
+          numinit
+        ];
+      };
     };
-  };
 in
 {
   z3_4_13 = common {

--- a/pkgs/applications/science/logic/z3/default.nix
+++ b/pkgs/applications/science/logic/z3/default.nix
@@ -101,6 +101,11 @@ let
           moveToOutput "lib/libz3java.${stdenv.hostPlatform.extensions.sharedLibrary}" "$java"
         '';
 
+      doInstallCheck = true;
+      installCheckPhase = ''
+        $out/bin/z3 -version 2>&1 | grep -F "Z3 version $version"
+      '';
+
       outputs =
         [
           "out"

--- a/pkgs/applications/science/logic/z3/lower-bound-typo.diff
+++ b/pkgs/applications/science/logic/z3/lower-bound-typo.diff
@@ -1,0 +1,13 @@
+diff --git a/src/@dir@/lp/column_info.h b/src/@dir@/lp/column_info.h
+index 1dc0c60..9cbeea6 100644
+--- a/src/@dir@/lp/column_info.h
++++ b/src/@dir@/lp/column_info.h
+@@ -47,7 +47,7 @@ public:
+             m_lower_bound_is_strict == c.m_lower_bound_is_strict &&
+             m_upper_bound_is_set == c.m_upper_bound_is_set&&
+             m_upper_bound_is_strict == c.m_upper_bound_is_strict&&
+-            (!m_lower_bound_is_set || m_lower_bound == c.m_low_bound) &&
++            (!m_lower_bound_is_set || m_lower_bound == c.m_lower_bound) &&
+             (!m_upper_bound_is_set || m_upper_bound == c.m_upper_bound) &&
+             m_cost == c.m_cost &&
+             m_is_fixed == c.m_is_fixed &&

--- a/pkgs/applications/science/logic/z3/tail-matrix.diff
+++ b/pkgs/applications/science/logic/z3/tail-matrix.diff
@@ -1,0 +1,12 @@
+diff --git a/src/@dir@/lp/tail_matrix.h b/src/@dir@/lp/tail_matrix.h
+index 2047e8c..c84340e 100644
+--- a/src/@dir@/lp/tail_matrix.h
++++ b/src/@dir@/lp/tail_matrix.h
+@@ -43,7 +43,6 @@ public:
+         const tail_matrix & m_A;
+         unsigned m_row;
+         ref_row(const tail_matrix& m, unsigned row): m_A(m), m_row(row) {}
+-        T operator[](unsigned j) const { return m_A.get_elem(m_row, j);}
+     };
+     ref_row operator[](unsigned i) const { return ref_row(*this, i);}
+ };

--- a/pkgs/development/haskell-modules/configuration-common.nix
+++ b/pkgs/development/haskell-modules/configuration-common.nix
@@ -1196,6 +1196,21 @@ self: super: {
     '';
   }) super.cryptol;
 
+  # Z3 removed aliases for boolean types in 4.12
+  inherit (
+    let
+      fixZ3 = appendConfigureFlags [
+        "--hsc2hs-option=-DZ3_Bool=bool"
+        "--hsc2hs-option=-DZ3_TRUE=true"
+        "--hsc2hs-option=-DZ3_FALSE=false"
+      ];
+    in
+    {
+      z3 = fixZ3 super.z3;
+      hz3 = fixZ3 super.hz3;
+    }
+  ) z3 hz3;
+
   # Tests try to invoke external process and process == 1.4
   grakn = dontCheck (doJailbreak super.grakn);
 

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -17303,12 +17303,12 @@ with pkgs;
   };
 
   inherit (callPackages ../applications/science/logic/z3 { python = python3; })
+    z3_4_13
     z3_4_12
     z3_4_11
-    z3_4_8;
-  inherit (callPackages ../applications/science/logic/z3 { python = python311; })
+    z3_4_8
     z3_4_8_5;
-  z3 = z3_4_8;
+  z3 = z3_4_13;
   z3-tptp = callPackage ../applications/science/logic/z3/tptp.nix { };
 
   tlaplus = callPackage ../applications/science/logic/tlaplus {

